migrations/datasets/4e2d7a28475b.py

```diff
@@ -0,0 +1,193 @@
+"""Migration script for ByDimensionsDatasetRecordStorageManagerUUID 2.0.0.
+
+Revision ID: 4e2d7a28475b
+Revises: 2101fbf51ad3
+Create Date: 2023-03-31 22:25:32.651155
+
+"""
+
+import datetime
+import itertools
+import logging
+from collections.abc import Callable
+from typing import Any
+
+import sqlalchemy as sa
+from alembic import context, op
+from lsst.daf.butler_migrate.butler_attributes import ButlerAttributes
+
+# revision identifiers, used by Alembic.
+revision = "4e2d7a28475b"
+down_revision = "2101fbf51ad3"
+branch_labels = None
+depends_on = None
+
+_LOG = logging.getLogger(f"lsst.daf.butler_migrate.{__name__}")
+
+# Manager name
+MANAGER = (
+    "lsst.daf.butler.registry.datasets.byDimensions._manager.ByDimensionsDatasetRecordStorageManagerUUID"
+)
+
+_EPOCH = datetime.datetime(1970, 1, 1)
+
+
+def _convert_to_ns(dt: datetime.datetime) -> int:
+    """Convert datetime to nanoseconds."""
+    microseconds = (dt - _EPOCH) // datetime.timedelta(microseconds=1)
+    return microseconds * 1000
+
+
+def _convert_to_datetime(nsec: int) -> datetime.datetime:
+    """Convert nanoseconds to datetime."""
+    microseconds = nsec // 1000
+    return datetime.datetime.utcfromtimestamp(microseconds / 1e6)
+
+
+def upgrade() -> None:
+    """Upgrade from version 1.0.0 to version 2.0.0.
+
+    Summary of changes:
+
+        - ingest_date column type changed from database-native TIMESTAMP
+          to nanoseconds-since-epoch in TAI scale.
+    """
+    bind = op.get_bind()
+    if bind.dialect.name == "postgresql":
+        # Keep it at miscrosecond resolution
+        _migrate_pg(
+            sa.BIGINT, None, "CAST(EXTRACT(EPOCH FROM ingest_date) * 1000000 AS BIGINT) * 1000", "2.0.0"
```

Comment:
I'm kind of amazed that this reproduces the Python calculation, but happy to trust that you've tested it.  A comment that it implicitly duplicates the 1970-01-01 constant in daf_butler's `time_utils.py` might be warranted, even though that is of course a pretty universal constant.

Reply:
Hmm, I did test it, but thinking about this again, I think it calculates something slightly different. In Python we convert astropy timestamps to TAI when before we send them to database, but SQL does not know about TAI, of course. This will produce UTC nanoseconds, so it's not exactly the same result. My general case is also doing everything in UTC, I guess I have to change that too.

Reply:
OK, I tried to do everything correctly without Postgres USING, but going through memory + temporary table and join but saving it as TAI instead of UTC. It works OK on small scale, but it is not going to work on large dataset tables. I'm running this on a copy of `main`, which has 143M records presently, and it's terribly slow, in 2 hours it's only about 15% done. Most of that time is spent in Python doing datetime->astropy->nanoseconds conversion for every record. I could probably speed up that a little bit, but I'm not very optimistic. If we could assume 37 seconds difference between UTC and TAI for all records then we could do things much faster in SQL.

Reply:
In some sense it's not *wrong* to do the time conversions in batches where you hardcode the TAI offset for each batch.

Reply:
This is hard to do in batches, this is a single ALTER TABLE command that needs to calculate new column value for each row in a table. I could probably write a single SQL expression with CASE and many branches that calculates TAI-UTC difference (as known today) but I doubt a reviewer can verify that it is correct (especially for the future). 🙂

Reply:
I think OODS data repositories are the only ones where `ingest_date` is not an arbitrary number (certainly nobody cares exactly when I ran the gen2to3 tools on all of the HSC data at NCSA, for example), and even OODS only cares about ~day accuracy.  So I'd vote for just assuming TAI=UTC (or maybe including 37s as a constant) in this script, while putting lots of warnings about this behavior in and around the migration script.

Reply:
OK, I updated the script to add/subtract 37 seconds in postgres USING expression. For non-postgres case I now use `TimeConverter` from `daf_butler` that is used by our `ddl` classes. I verified that things are now working correctly for both PostgreSQL and SQLite. Optimized postgres case needs less than 5 minutes on `main` repo copy, compared to ~10 hours for non-optimized case.

@TallJimbo, could you check my latest commit [80d1c59](https://github.com/lsst-dm/daf_butler_migrate/pull/21/commits/80d1c592123f92541467a653a10fe51929c70a89)?

Reply:
Re-reviewed, looks good!